### PR TITLE
OpenAI and Huggingface Providers

### PR DESCRIPTION
**Pull request by Spartee**
Providers for creating embeddings with OpenAI and Hugginface.

Includes
- base class
- simple integration testing
- async methods for clients that support it.

### CODE REVIEW

File: redisvl/providers/__init__.py [hunk 1]
```diff
@@ -0,0 +1,9 @@
+
+from redisvl.providers.openai import OpenAIProvider
```
Comment (tylerhutcherson) [inline redisvl/providers/__init__.py:2]: should we consider organizing it one layer deeper, under `embeddings`? that way its clear whether its for Embedding creation or something else (future proofing)

File: redisvl/providers/base.py [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+    async def aembed_many(
+        self, inputs: List[str], preprocess: callable = None, chunk_size: int = 1000
+    ) -> List[float]:
+        raise NotImplementedError
+
+    async def aembed(self, emb_input: str, preprocess: callable = None) -> List[float]:
+        raise NotImplementedError
+
+    def batchify(self, seq: list, size: int, preprocess: Callable = None):
```
Comment (tylerhutcherson) [inline redisvl/providers/base.py:38]: love it

File: redisvl/providers/huggingface.py [hunk 3]
```diff
@@ -0,0 +1,29 @@
+from typing import Dict, List, Optional
+
+from sentence_transformers import SentenceTransformer
```
Comment (tylerhutcherson) [inline redisvl/providers/huggingface.py:3]: we could consider importing this scoped within the class -- and then we can prevent folks from having to install SentenceTransformers (torch, etc) if they don't plan to need it?

File: redisvl/providers/openai.py [hunk 4]
```diff
@@ -0,0 +1,48 @@
+from typing import Dict, List, Optional
+
+import openai
```
Comment (tylerhutcherson) [inline redisvl/providers/openai.py:3]: same point as above?

File: redisvl/providers/openai.py [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+    def __init__(self, model: str, api_config: Optional[Dict] = None):
+        dims = 1536
+        super().__init__(model, dims, api_config)
+        if not api_config:
+            raise ValueError("OpenAI API key is required in api_config")
+
+        openai.api_key = api_config.get("api_key", None)
+        self._model_client = openai.Embedding
+
+    def embed_many(
```
Comment (tylerhutcherson) [inline redisvl/providers/openai.py:18]: should we look to make this immune to rate limit issues? One pattern I have seen is to use `tenacity` and their exponential backoff decorator.

**Review by tylerhutcherson — APPROVED** [review]
Left some suggestions but looks great all around!

File: redisvl/providers/huggingface.py (hunk 3 above)
Comment (Spartee) [reply redisvl/providers/huggingface.py:3]: It's the default for LLMCache... should we not assume that? I do like the scoped gaurd approach.

File: redisvl/providers/openai.py (hunk 5 above)
Comment (Spartee) [reply redisvl/providers/openai.py:18]: @tylerhutcherson can you point me to or write a code sample for this?

File: redisvl/providers/__init__.py (hunk 1 above)
Comment (Spartee) [reply redisvl/providers/__init__.py:2]: That's a long import but I see what you mean. A couple options

```python
from redisvl.api.embedding.openai
from redisvl.embedding.openai import OpenAIEmbedding
from redisvl.model.embedding.openai import ...

```

others?

**Comment by Spartee** [comment]
addressing changes in Next PR

**Pull request merged**